tsl/src/fdw/deparse.c

```diff
@@ -834,6 +834,128 @@ deparseSelectStmtForRel(StringInfo buf, PlannerInfo *root, RelOptInfo *rel, List
 	deparseLockingClause(&context);
 }
 
+/*
+ * Construct "SELECT DISTINCT target_list" or "SELECT DISTINCT ON (col1, col..)
+ * target_list" statement to push down the DISTINCT clause to the remote side.
+ *
+ * We only allow references to basic "Vars" or constants in the DISTINCT exprs
+ *
+ * So, "SELECT DISTINCT col1" is fine but "SELECT DISTINCT 2*col1" is not.
+ *
+ * "SELECT DISTINCT col1, 'const1', NULL, col2" which is a mix of column
+ * references and constants is also supported. Everything else is not supported.
+ *
+ * It should be noted that "SELECT DISTINCT col1, col2" will return the same
+ * set of values as "SELECT DISTINCT col2, col1". So nothing additional needs
+ * to be done here as the upper projection will take care of any ordering
+ * between the attributes.
+ *
+ * We also explicitly deparse the distinctClause entries only for the
+ * "DISTINCT ON (col..)" case. For regular DISTINCT the targetlist
+ * deparsing which happens later is good enough
+ */
+static void
+deparseDistinctClause(StringInfo buf, deparse_expr_cxt *context)
+{
+	PlannerInfo *root = context->root;
+	Query *query = root->parse;
+	ListCell *l, *dc_l;
+	bool first = true;
+	bool dc_pushdown_safe = true;
+	Index varno;
+
+	if (query->distinctClause == NIL)
+		return;
+
+	foreach (l, query->distinctClause)
+	{
+		SortGroupClause *sgc = lfirst_node(SortGroupClause, l);
+		TargetEntry *tle = get_sortgroupclause_tle(sgc, query->targetList);
+
+		/*
+		 * We only send basic attributes to the remote side. So we can
+		 * pushdown DISTINCT only if the tle is a simple one
+		 * referring to the "Var" directly. Also all varno entries
+		 * need to point to the same relid.
+		 *
+		 * Also handle "DISTINCT col1, CONST1, NULL" types cases
+		 */
+		if (IsA(tle->expr, Var))
+		{
+			Var *var = (Var *) tle->expr;
```

Comment:
```suggestion
			Var *var = castNode(Var,tle->expr);
```
castNode has the added benefit of adding an Assert that its right type

Reply:
We do an `IsA` to check if it's a var just above. However `castNode` use is much cleaner. Done.